### PR TITLE
SD: Implementing directional cosine matrices and section properties for rectangular members

### PR DESCRIPTION
**Pull request by samuel-ramsahoye**
This PR is ready to be merged.

Compilation has only been experimented with via compilation of the SubDyn driver ("make subdyn_driver"). The physical correctness will be checked soon via some canonical tests.

No new tests have been added yet for this new functionality - existing tests may fail.

**Feature or improvement description**
SubDyn primary input files allow the user to specify "X" section properties and member cosines as tables but currently these inputs are not used. This PR aims to correct that.

This update should be backwards compatible with existing SubDyn primary input files. If the COSMID column is missing from the member specification, SubDyn's default member cosine convention is used. Moreover, the user can partially specify member cosines by specifying -1 in place of COSMID for the relevant rows: -1 is the fill value for the entire column if it's left unspecified.

"X" section properties for a member are used if MType=4. Whilst MType=4 members can be refined, the property ID at either joint must be the same. If not, a fatal exception occurs and an appropriate message is displayed. The definition of "X" section properties hasn't changed from previous implementations.

User-specified cosines are a direct replacement for the GetDirCos call in SD_FEM.f90 and so they are in fact the transpose of what would normally be considered the member cosine matrix. At the moment, no checks are in place to ensure the correctness of user-defined cosines.

**Related issue, if one exists**
https://github.com/OpenFAST/openfast/issues/1404

**Impacted areas of the software**
This PR relates to the SubDyn module only.

Changes predominantly relate to SubDyn.f90, SD_FEM.f90 and FEM.f90.

**Additional supporting information**
The "./subdyn_cosine_arbitrary_example" has been used during initial testing. This code has only be experimented with via the SubDyn driver and not via the OpenFAST glue-code and compilation has only been checked via "make subdyn_driver".

**Test results, if applicable**
Tests will be added to this draft PR is due course.

### CODE REVIEW

**Comment by samuel-ramsahoye** [comment]
@ebranlard Please see the draft PR for the member cosines and X section properties issue.

**Comment by samuel-ramsahoye** [comment]
@bjonkman Thanks very much Bonnie!

**Comment by samuel-ramsahoye** [comment]
@ebranlard I've found a couple of issues with my PR in my regression tests, the main one being that I don't think the stiffness matrices are actually being evaluated for members with X sections yet. I know what I need to modify in the code to fix this.

For the testing, to start with what I'm doing is taking existing SubDyn models and re-writing them by specifying member cosines and properties via X sections and checking the results are still the same.

**Comment by samuel-ramsahoye** [comment]
@ebranlard There's an issue with the user-specified cosines in this PR. I set up a simple angled cantilever example, the modes of which should be the same irrespective of whether or not the member cosines are specified by the user or calculated by GetDirCos. I populated the COSM information by writing it out from SubDyn. The modal frequencies are the same but the displacements are different, except for the first mode. The relevant input files for this case are in the ./f_subdyn_test/02 folder if you want to take a look. You can reconstruct this behaviour by toggling COSMID value of the single member between -1 and 1.

File: modules/subdyn/src/SD_FEM.f90 [hunk 1]
```diff
@@ -397,7 +402,7 @@ SUBROUTINE SD_ReIndex_CreateNodesAndElems(Init,p, ErrStat, ErrMsg)
          ! Test that the two properties match for non-beam 
          if (mType/=idMemberBeam) then
              if (Init%Members(iMem, iMProp)/=Init%Members(iMem, iMProp+1)) then
-                call Fatal('Properties should be the same at each node for non-beam members. Check member with ID: '//TRIM(Num2LStr(Init%Members(iMem,1))))
+                call Fatal('Property IDs should be the same at each joint for members with X section properties. Check member with ID: '//TRIM(Num2LStr(Init%Members(iMem,1))))
```
Comment (ebranlard) [inline modules/subdyn/src/SD_FEM.f90:405]: It seems that this check was previously for anything that is not a beam (Rigid, Cable), and now we restrict it as well for anisotropic beams.  Shouldn't the text "with X section properties" be "for anisotropic beams, rigid links, and cables" maybe?

File: modules/subdyn/src/SD_FEM.f90 [hunk 2]
```diff
@@ -436,11 +447,13 @@ SUBROUTINE SD_Discrt(Init,p, ErrStat, ErrMsg)
    INTEGER                       :: NNE      ! number of nodes per element
    INTEGER                       :: MaxNProp
    REAL(ReKi), ALLOCATABLE       :: TempProps(:, :)
+   REAL(ReKi), ALLOCATABLE       :: TempPropsX(:, :)
    INTEGER, ALLOCATABLE          :: TempMembers(:, :)
    INTEGER                       :: knode, kelem, kprop, nprop
+   INTEGER                       :: COSM_ind
```
Comment (ebranlard) [inline modules/subdyn/src/SD_FEM.f90:453]: I would maybe use `iDirCos` and add a comment  `!< index of directional cosine matrix (-1 if not used)`

File: modules/subdyn/src/SD_FEM.f90 [hunk 3]
```diff
@@ -563,19 +589,28 @@ SUBROUTINE SD_Discrt(Init,p, ErrStat, ErrMsg)
...
+            dd = ( d2 - d1 )/Init%NDiv
+            dt = ( t2 - t1 )/Init%NDiv
+
              ! If both dd and dt are 0, no interpolation is needed, and we can use the same property set for new nodes/elements. otherwise we'll have to create new properties for each new node
-          CreateNewProp = .NOT. ( EqualRealNos( dd , 0.0_ReKi ) .AND.  EqualRealNos( dt , 0.0_ReKi ) )  
-          
+           
+            CreateNewProp = .NOT. ( EqualRealNos( dd , 0.0_ReKi ) .AND.  EqualRealNos( dt , 0.0_ReKi ) ) 
+
+          else
```
Comment (ebranlard) [inline modules/subdyn/src/SD_FEM.f90:608]: we might as well be explicit here and use `else if (eType == idMemberX)` for readability

File: modules/subdyn/src/SD_FEM.f90 [hunk 4]
```diff
@@ -770,10 +807,12 @@ SUBROUTINE SetElementProperties(Init, p, ErrStat, ErrMsg)
    INTEGER                  :: I
    INTEGER                  :: N1, N2     ! starting node and ending node in the element
    INTEGER                  :: P1, P2     ! property set numbers for starting and ending nodes
+   INTEGER                  :: COSM_ind
    REAL(ReKi)               :: D1, D2, t1, t2, E, G, rho ! properties of a section
    REAL(FEKi)               :: DirCos(3, 3)              ! direction cosine matrices
+   REAL(FEKi)                       :: Dx_, Dy_, Dz_! distances between nodes
```
Comment (ebranlard) [inline modules/subdyn/src/SD_FEM.f90:813]: Alignment appears to be off here. Make sure you don't use "Tab" in the code and only spaces. 

Anyway, I'm guessing these variables are temporary variables for debug and won't be needed.

File: modules/subdyn/src/SD_FEM.f90 [hunk 5]
```diff
@@ -793,11 +832,44 @@ SUBROUTINE SetElementProperties(Init, p, ErrStat, ErrMsg)
...
+         DirCos(1, 1) =  Init%COSMs(COSM_ind, 2)
+         DirCos(1, 2) =  Init%COSMs(COSM_ind, 3)
+         DirCos(1, 3) =  Init%COSMs(COSM_ind, 4)
+         DirCos(2, 1) =  Init%COSMs(COSM_ind, 5)
+         DirCos(2, 2) =  Init%COSMs(COSM_ind, 6)
+         DirCos(2, 3) =  Init%COSMs(COSM_ind, 7)
+         DirCos(3, 1) =  Init%COSMs(COSM_ind, 8)
+         DirCos(3, 2) =  Init%COSMs(COSM_ind, 9)
+         DirCos(3, 3) =  Init%COSMs(COSM_ind, 10)
+         DirCos = TRANSPOSE(DirCos)
```
Comment (ebranlard) [inline modules/subdyn/src/SD_FEM.f90:855]: Yes, I think we might need this transpose. I think SubDyn use an opposite convention as the rest of OpenFAST. The DirCos of OpenFAST are always from inertial to body. But SubDyn uses the transpose. Please double check. We should make sure that the documentation of OpenFAST is clear on what to put in the DirCos table

File: modules/subdyn/src/SD_FEM.f90 [hunk 6]
```diff
@@ -853,7 +926,39 @@ SUBROUTINE SetElementProperties(Init, p, ErrStat, ErrMsg)
          p%ElemProps(i)%Iyy    = Iyy
          p%ElemProps(i)%Jzz    = Jzz
          p%ElemProps(i)%Shear  = Shear
-         p%ElemProps(i)%kappa  = kappa
+         p%ElemProps(i)%kappa_x  = kappa
+         p%ElemProps(i)%kappa_y  = kappa
```
Comment (ebranlard) [inline modules/subdyn/src/SD_FEM.f90:930]: The code appears to be inconsistent whether we use `kappa` or `Kappa` (not your fault). This pull request is our chance to change that everywhere since you are changing this variable name anyway. Let's use upper case `Kappa` (`Kappa_x` and `Kappa_y`)  to be consistent with the other fields already defined in the registry. 

Please only change the variable `kappa`, not the other ones (as it would create noise in the pull request).

File: modules/subdyn/src/SubDyn.f90 [hunk 7]
```diff
@@ -1152,13 +1152,45 @@ SUBROUTINE SD_Input(SDInputFile, Init, p, ErrStat,ErrMsg)
 if (LegacyFormat) then
    nColumns = 5
    Init%Members(:,iMType) = idMemberBeam ! Important, in legacy all members are beams
+   Init%Members(:,7) = -1
 else
    nColumns = MembersCol
 endif
-DO I = 1, p%NMembers
+
```
Comment (ebranlard) [inline modules/subdyn/src/SubDyn.f90:1159]: We can add a fatal statement here if the number of member is zero. (which indeed woulc be an invalid SubDyn model). 
Then we can be sure that we read the "first line" of the Member table below, because it should always exists.

File: modules/subdyn/src/SubDyn.f90 [hunk 8]
```diff
@@ -1152,13 +1152,45 @@ SUBROUTINE SD_Input(SDInputFile, Init, p, ErrStat,ErrMsg)
 if (LegacyFormat) then
```
Comment (ebranlard) [inline modules/subdyn/src/SubDyn.f90:1152]: NOTE: the LegacyFormat has been here for a couple of release, it's likely time to remove it. Now we can introduce a new legacy format regarding your new changes...

File: modules/subdyn/src/SubDyn.f90 [hunk 9]
```diff
@@ -1152,13 +1152,45 @@ SUBROUTINE SD_Input(SDInputFile, Init, p, ErrStat,ErrMsg)
...
-DO I = 1, p%NMembers
+
+nColumns=MembersCol
+CALL AllocAry(StrArray, nColumns, 'StrArray',ErrStat2,ErrMsg2); if (Failed()) return 
+READ(UnIn, FMT='(A)', IOSTAT=ErrStat2) Line  ; ErrMsg2='First line of members array'; if (Failed()) return
+CALL ReadCAryFromStr ( Line, StrArray, nColumns, 'Members', 'First line of members array', ErrStat2, ErrMsg2 )
+if (ErrStat2/=0) then
+   ! We try with 4 columns (legacy format)
+   nColumns = MembersCol-1
+   deallocate(StrArray)
```
Comment (ebranlard) [inline modules/subdyn/src/SubDyn.f90:1167]: always add an if statement: `if (allocated(StrArray)) deallocate(StrArray)`  to be safe

File: modules/subdyn/src/SubDyn.f90 [hunk 10]
```diff
@@ -1152,13 +1152,45 @@ SUBROUTINE SD_Input(SDInputFile, Init, p, ErrStat,ErrMsg)
...
+! Extract fields from first line
+DO I = 1, nColumns
+   bInteger = is_integer(StrArray(I), Init%Members(1,I)) ! Convert from string to float
+   if (.not.bInteger) then
+      CALL Fatal(' Error in file "'//TRIM(SDInputFile)//'": Non integer character found in Member line. Problematic line: "'//trim(Line)//'"')
+      return
+   endif
+ENDDO
+
+deallocate(StrArray)
```
Comment (ebranlard) [inline modules/subdyn/src/SubDyn.f90:1182]: add if statement

File: modules/subdyn/src/SubDyn.f90 [hunk 11]
```diff
@@ -1152,13 +1152,45 @@ SUBROUTINE SD_Input(SDInputFile, Init, p, ErrStat,ErrMsg)
...
    nColumns = MembersCol
 endif
-DO I = 1, p%NMembers
+
+nColumns=MembersCol
+CALL AllocAry(StrArray, nColumns, 'StrArray',ErrStat2,ErrMsg2); if (Failed()) return 
+READ(UnIn, FMT='(A)', IOSTAT=ErrStat2) Line  ; ErrMsg2='First line of members array'; if (Failed()) return
+CALL ReadCAryFromStr ( Line, StrArray, nColumns, 'Members', 'First line of members array', ErrStat2, ErrMsg2 )
+if (ErrStat2/=0) then
+   ! We try with 4 columns (legacy format)
```
Comment (ebranlard) [inline modules/subdyn/src/SubDyn.f90:1165]: "We try with one column less (legacy format)"

File: modules/subdyn/src/SubDyn.f90 [hunk 12]
```diff
@@ -1152,13 +1152,45 @@ SUBROUTINE SD_Input(SDInputFile, Init, p, ErrStat,ErrMsg)
...
+   nColumns = MembersCol-1
+   deallocate(StrArray)
+   CALL AllocAry(StrArray, nColumns, 'StrArray',ErrStat2,ErrMsg2); if (Failed()) return 
+   CALL ReadCAryFromStr ( Line, StrArray, nColumns, 'Members', 'First line of members array', ErrStat2, ErrMsg2 ); if(Failed()) return
+   call LegacyWarning('Member table contains 6 columns instead of 7, using default member cosines.')
```
Comment (ebranlard) [inline modules/subdyn/src/SubDyn.f90:1170]: Suggest modifying to: ", using default member directional cosines ID (-1) for all members. The directional cosine will be computed based on the member nodes for all members. "

File: modules/subdyn/src/SubDyn.f90 [hunk 13]
```diff
@@ -1152,13 +1152,45 @@ SUBROUTINE SD_Input(SDInputFile, Init, p, ErrStat,ErrMsg)
...
+   endif
+ENDDO
+
+deallocate(StrArray)
+! ! Read remaining lines
+DO I = 2, p%NMembers
    CALL ReadAry( UnIn, SDInputFile, Dummy_IntAry, nColumns, 'Members line '//Num2LStr(I), 'Member number and connectivity ', ErrStat2,ErrMsg2, UnEc); if(Failed()) return
-   Init%Members(I,1:nColumns) = Dummy_IntAry(1:nColumns)
-ENDDO   
+   Init%Members(I,1:(nColumns)) = Dummy_IntAry(1:(nColumns))
```
Comment (ebranlard) [inline modules/subdyn/src/SubDyn.f90:1186]: parentheses can be removed again

File: f_subdyn_test/log.txt [hunk 14]
```diff
@@ -0,0 +1,2 @@
+1. vertical cantilever
```
Comment (ebranlard) [inline f_subdyn_test/log.txt:1]: In the r-test, use README.md for these kind of files if you can.

**Review by ebranlard — COMMENTED** [review]
Thanks for doing that, the changes looks really good. I mostly have a couple of comments regarding some conventions that I would personally prefer. 

One thing I couldn't comment on but might be important:
- I would recommend setting COSMs (in the registry) as R8Ki and not ReKi. Maybe this is the reason why you see different results
Also:
- I would recommend using another test case, where you keep the regular Beam type (3) and specify the directional cosine. (to avoid testing both the directional cosine feature and the anisotropic member feature). 

Some important points to think about (and probably discuss with @jjonkman):
- Do we agree that "anisotropic" (Aniso for short) is better than "X" ?
- do we want to allow for division of anisotropic members?
- Should we average the properties on both ends of the anisotropic member? (We need to think again of what is done in SubDyn when there are different properties on both node ends).
- We need to make sure the convention for the directional cosine inputs is well documented (given the internal transpose required).

File: f_subdyn_test/01/driver.inp [hunk 15]
```diff
@@ -0,0 +1,26 @@
+SubDyn Driver file for stand-alone applications
```
[the remark this answers is not in the document]
Comment (samuel-ramsahoye) [reply f_subdyn_test/01/driver.inp:1]: Yes I completely agree - currently just there incidentally during development.

File: modules/subdyn/src/SD_FEM.f90 [hunk 16]
```diff
@@ -41,6 +41,7 @@ MODULE SD_FEM
   INTEGER(IntKi),   PARAMETER  :: CMassCol        = 11                    ! Number of columns in Concentrated Mass (CMJointID,JMass,JMXX,JMYY,JMZZ, Optional:JMXY,JMXZ,JMYZ,CGX,CGY,CGZ)
   ! Indices in Members table
   INTEGER(IntKi),   PARAMETER  :: iMType= 6 ! Index in Members table where the type is stored
+  INTEGER(IntKi),   PARAMETER  :: iMCOSMID = 7 ! Index in Members table where the type is stored
```
[the remark this answers is not in the document]
Comment (samuel-ramsahoye) [reply modules/subdyn/src/SD_FEM.f90:44]: Sure, I can make some other similar changes too.

File: modules/subdyn/src/SD_FEM.f90 (hunk 1 above)
Comment (samuel-ramsahoye) [reply modules/subdyn/src/SD_FEM.f90:405]: Yes I didn’t realise this check was repeated twice as you have noted elsewhere.

File: modules/subdyn/src/SD_FEM.f90 (hunk 3 above)
Comment (samuel-ramsahoye) [reply modules/subdyn/src/SD_FEM.f90:608]: Agreed

File: modules/subdyn/src/SD_FEM.f90 (hunk 5 above)
Comment (samuel-ramsahoye) [reply modules/subdyn/src/SD_FEM.f90:855]: Don’t worry the exact convention will be qualified in the tests rather than being assumed.

File: modules/subdyn/src/SD_FEM.f90 (hunk 6 above)
Comment (samuel-ramsahoye) [reply modules/subdyn/src/SD_FEM.f90:930]: Sure

File: modules/subdyn/src/SubDyn.f90 (hunk 7 above)
Comment (samuel-ramsahoye) [reply modules/subdyn/src/SubDyn.f90:1159]: Agreed

File: modules/subdyn/src/SubDyn.f90 (hunk 8 above)
Comment (samuel-ramsahoye) [reply modules/subdyn/src/SubDyn.f90:1152]: Agreed

File: modules/subdyn/src/SubDyn.f90 (hunk 10 above)
Comment (samuel-ramsahoye) [reply modules/subdyn/src/SubDyn.f90:1182]: Sure

File: modules/subdyn/src/SubDyn.f90 (hunk 9 above)
Comment (samuel-ramsahoye) [reply modules/subdyn/src/SubDyn.f90:1167]: Sure

File: f_subdyn_test/log.txt (hunk 14 above)
Comment (samuel-ramsahoye) [reply f_subdyn_test/log.txt:1]: Yes of course, will formalise with README files once I’m completely satisfied with the results

File: modules/subdyn/src/SubDyn.f90 (hunk 8 above)
Comment (samuel-ramsahoye) [reply modules/subdyn/src/SubDyn.f90:1152]: @ebranlard It was also problematic because LegacyFormat was previously inferred from the joint parsing rather than the member parsing. In this instance, the member type will have been excluded and assumed to be one even if the user desired a rigid or cable.

**Comment by samuel-ramsahoye** [comment]
@ebranlard Thanks very much for your comments. I’ve made the majority of the changes you suggested. I believe the outstanding ones are conditional on @jjonkman’s opinion.

Your suggestion of increasing the number of bits used to store the direction cosine information worked when additionally used in conjunction with writing and reading the member cosines to a much higher precision. The precision with which the user needs to specify these cosines is very high in order for the results to match “perfectly”. This makes me think that specification via a rotational angle about the direction vector, rather than 9 high precision elements of a matrix, might be better in retrospect.

To summarise, results match when both:
1. The user directly specifies CHS properties via equivalent X section properties
2. The user specifies member direction cosines rather than using the internal evaluation in SubDyn

The next few steps for me (aside from any further recommendations from yourself):

1. Formalise a set of canonical tests that separate the checking of each of these two new features independently as well as together
2. I’d like to review the changes I’ve made again now all the functionality appears to be working as expected

Cheers,
Sam

**Comment by ebranlard** [comment]
Thanks a lot for doing the changes! You'll see that the tests are failing in single precision, because FINLOCI (in FEM.f90) was written for ReKi or Int. The easiest is probably to add another interface to FINDLOCI (for SiKi) and change the existing one from ReKi to R8Ki. 

All rotations in OpenFAST are stored as double precision as these parameters can indeed be quite sensitive. We could input the directional cosines as 3 Euler angles or 4 Euler parameters in the input file (preferable to only 3 parameters), but I would expect those to also have to be given with high precision.

**Comment by samuel-ramsahoye** [comment]
@ebranlard For our particular use case, it's fine to assume the direction cosine of the local element Z-axis is the element direction vector. The orientation of the local element X and Y axes can then be determined by a single rotational angle about the local element Z-axis. I think this might have been what you originally suggested. This could be a member (and element) property rather than (a column in the input file in the member table) and we could then get rid of the direction cosines table.

We could probably do with some error handling on user-specified cosines: I'll try to add in the next few days.

Cheers,
Sam

**Comment by ebranlard** [comment]
Hi @samuel-ramsahoye , yes, the rotation around the z axis was indeed what I had in mind for ease of input by the user. The user still have to think about/figure out where the "x" axis is placed by default by SubDyn. I would probably hold on the implementation for now until we take a decision, or at least not remove the direction cosine table.

**Comment by samuel-ramsahoye** [comment]
@ebranlard Yes I’ll hold off. Does NREL provide a means to visualise OpenFAST models? It would be nice if the user could visually inspect the orientation of the local-element axes in SubDyn, amongst other things from other modules too. If not I can write something in Python.

**Comment by ebranlard** [comment]
I hope to get back to you in a week or so for the remaining decisions. 

Regarding visualization of the orientation, I've never done it, but my colleague @andrew-platt  has. The only available option at the moment is to output the meshes to VTK using `WrVTK` and `VTK_Type = 2` (for instance). Then in ParaView, you can use 3D glyph arrows to visualize each component of the orientation matrix (which is basically your frame vectors):
Screen Shot 2023-01-25 at 10 46 26 AM (1)

**Comment by andrew-platt** [comment]
Following on what @ebranlard mentioned, here is an example of what visualizing the orientations can look like (X = red, Y = yellow, Z = green).  This example is from visualization of the blade root orientations from two simulations (lighter arrows are the reference simulation).

Screen Shot 2023-01-25 at 10 56 35 AM

**Comment by ebranlard** [comment]
Hi @samuel-ramsahoye 
Thanks again for your work and your patience. 
I discussed some of our remaining questions with @jjonkman. In the meantime, I also realized that the "X" members were still isotropic elements. 
  
Here's what we agreed on:
- The "regular Beam" could be renamed "BeamCirc" (for circular cross section). The "X" beams could be renamed BeamRect  (for rectangular cross section). It's totally fine if you don't feel like tackling this renaming in the code. 
- The BeamRect can be allowed to be subdivided, using linear interpolation for geometrical quantities. If you don't have the time to implement it, you can leave a comment in the code and in the documentation for now. 
- Material properties such as Young modulus, Shear modulus, should be kept constant along a member, so it's important that we have these checks. 
- We can keep the directional cosine method that you have implemented, even if it's tedious for the user to provide the 9 values. In the future, we can implement an "automatic" method for the orientation of these members, where the "y" axis (for instance) is always in the horizontal plane (and the z axis always along the member nodes). 

Do you think you have the time to address some of the remaining comments? Then we can work on setting up a test case and merge this pull request.
Thanks a lot!

**Comment by Sam-Ramsahoye** [comment]
@ebranlard Thanks for getting in touch and good to hear from you.

If you wouldn't mind, two things currently aren't clear to me:

1. the changes I've made aren't specific to rectangular beams. Following the conventions in the current input file, non-circular sections in this PR are currently defined in the same way as they've always been, with no distinction between members of any type, be they rectangular, channel etc as the user inputs the geometric properties relevant to the FEM model directly (areas, shear areas, second moments of area etc).
2. ", I also realized that the "X" members were still isotropic elements." - I'm not sure what you mean by this. As far as I can see, in my PR, the user can define section y, z properties independently and they're passed into the element stiffness definition.

Is your intention to ultimately let users only define circular and rectangular sections? I haven't be a part of those conversations so I might be slightly out of the loop.

I'm still happy to help with any of the remaining work.

Cheers,
Sam

**Comment by ebranlard** [comment]
Hi @Sam-Ramsahoye 

I think we assumed that the elements would mostly be rectangular, but that indeed that doesn't have to be the case. Then NonCirc" or "Arb" (for arbitrary) would be better. 

Sorry for the confusion about isotropic. I was reading the input file of SubDyn and seeing that the "X-member" section was intended for "isotropic" material, which I believe is isotropic in the sense that the material is the same throughout the member (same density, E, and G), but indeed the geometrical properties are not axisymmetric and can vary along the span of the member. 

Let me know if that makes more sense!
Thanks again for all your work,

Emmanuel

**Comment by Sam-Ramsahoye** [comment]
@ebranlard No worries, thanks for clarifying.

I can look at making some of those changes over the next couple of days. The only thing I might not have time to do is write in the interpolation of NonCirc properties. Currently, those elements are subdivided because the simplification that the start and end properties must be the same makes it simple.

**Comment by ebranlard** [comment]
Hi @verlivkra, did the method mentioned by @andrew-platt  not work for you? (I have not tried it..)

**Comment by ebranlard** [comment]
Hi @Sam-Ramsahoye, have you made any progress on the remaining implementation detail? We would like to merge this pull request in the next couple of weeks. Thanks!

**Comment by Sam-Ramsahoye** [comment]
@ebranlard  Apologies for my delay in getting back to you. I've been very busy recently but planning to look at this early next week.

Apologies for the delay again.

Sam

**Comment by verlivkra** [comment]
@ebranlard as far as I can see, no SubDyn-members are visualized when generating vtk-files from OpenFAST. I don't see how 3D-glyphs would then be available for the SubDyn-members. What I saw from @andrew-platt 's comment was only an example on how this could look for the blades, which I assume are not SubDyn-members. Did I miss something here? 
If there is a better place to continue this discussion, please let me know.

**Comment by ebranlard** [comment]
Hi @samuel-ramsahoye 
Do you have time to wrap up this pull request? We are hoping to put that in our 3.5.0 release in the coming weeks. If you don't have time, can you grant me access to your branch? (add me as a collaborator) 
Could you also add a zip file in a post here with the small test cases you had put together? 
Thanks!

**Comment by samuel-ramsahoye** [comment]
Hi @ebranlard,

Really sorry I’ve been so slow. Unfortunately, I haven’t had time to have a look at this recently and a jacket project I had planned on studying in OpenFAST, where the functionality would have been useful, got postponed until May.

I’ll add you as a contributor and try and gather the (very) limited test cases I’ve put together.

Cheers,
Sam

File: modules/subdyn/src/SD_FEM.f90 [hunk 17]
```diff
@@ -455,16 +469,17 @@ SUBROUTINE SD_Discrt(Init,p, ErrStat, ErrMsg)
...
+   nMemberRigid    = count(Init%Members(:,iMType) == idMemberRigid)
+   nMemberBeamArb  = count(Init%Members(:,iMType) == idMemberBeamArb)
+   Init%NElem = (nMemberBeamCirc + nMemberBeamArb)*Init%NDiv + nMemberCable + nMemberRigid  ! NOTE: only Beams are divided
+   IF ( (nMemberBeamCirc+nMemberRigid+nMemberCable+nMemberBeamArb) /= size(Init%Members,1)) then
       CALL Fatal(' Member list contains an element which is not a beam, a cable or a rigid link'); return
    ENDIF
 
    ! Total number of nodes - Depends on division and number of nodes per element
-   p%nNodes = Init%NJoints + ( Init%NDiv - 1 )*nMemberBeam
+   p%nNodes = Init%NJoints + ( Init%NDiv - 1 )*(nMemberBeamCirc + nMemberBeamArb)
```
Comment (ebranlard) [inline modules/subdyn/src/SD_FEM.f90]: This might need to be changed because Arb beam are not divided at the moment.

File: modules/subdyn/src/SD_FEM.f90 [hunk 18]
```diff
@@ -853,13 +919,87 @@ SUBROUTINE SetElementProperties(Init, p, ErrStat, ErrMsg)
...
+         print*,'p%ElemProps(i)%Kappa_y',p%ElemProps(i)%Kappa_y*A
+         print*,''
+
+         print*,'p%ElemProps(i)%Ixx    ',p%ElemProps(i)%Ixx    
+         print*,'p%ElemProps(i)%Iyy    ',p%ElemProps(i)%Iyy    
+         print*,'p%ElemProps(i)%Jzz    ',p%ElemProps(i)%Jzz    
+         print*,''
+
+         print*,'p%ElemProps(i)%Shear  ',p%ElemProps(i)%Shear  
+         print*,'p%ElemProps(i)%D      ',p%ElemProps(i)%D
```
Comment (ebranlard) [inline modules/subdyn/src/SD_FEM.f90]: remove print statements

**Review by ebranlard — APPROVED** [review]

**Comment by Sam-Ramsahoye** [comment]
@ebranlard Apologies I didn't get around to putting together the tests for you, I've been very busy.

I've got a couple of uses for this functionality coming up so I'm going to collate the tests then - some additional checks that I've been performing seem to compare well.

Would it be cleaner to compare expected and calculated stiffness matrices rather than time-series outputs from quasi-static sims?

I'm aware this has now been merged but I've found a small bug in this update. If you don't specify any beams with regular circular sections, an error is thrown. This is because currently when the p%ElemProps is set for eType==idMemberBeamArb, D1 and D2 are referenced but undefined. In any case, these are erroneous references and should probably replaced with A**0.5 or some other equivalent diameter.

**Comment by ebranlard** [comment]
Thanks for looking into it! Hopefully #1531 fixes this issue.
